### ISSUE TITLE
When exactly does 429 Too Many Requests occur?

### ISSUE THREAD

**Issue by Deleora** [instaloader/instaloader#1226]
Hi, 
So I wrote a Python script which would basically find out if there were any new story(s) uploaded or any story(s) deleted and in either of the case it would attempt to download the stories of the profile. I'm running this script in background and the script executes every minute. Based on [#673](https://github.com/instaloader/instaloader/issues/673#issuecomment-645215066) we can have minimum 20 seconds gap between making requests and I had set 60 seconds interval.
However, still I got some error (probably 429) which I can't tell exactly as the script was running in background so no console to log error. I then ran just the downloading story command in CMD and there I saw the following error:

Command: `instaloader --stories-only --no-posts --no-compress-json --login <my_username> <target_profile>`
Log:
```
Loaded session from E:\Instaloader\session-files\.instaloader-hp\session-<my_username>.
Logged in as <my_username>.
JSON Query to <target_profile>/feed/: 429 Too Many Requests [retrying; skip with ^C]
Number of requests within last 10/11/20/22/30/60 minutes grouped by type:
   d6f4427fbe92d846298cf<truncated>:    1    1    1    1    1    1
 *                            other:    1    1    1    1    1    1
Instagram responded with HTTP error "429 - Too Many Requests". Please
do not run multiple instances of Instaloader in parallel or within
short sequence. Also, do not use any Instagram App while Instaloader
is running.
The request will be retried in 665 seconds, at 15:42.
```

So can anyone please help me understand the exact reason why this happened?

I got to know about this when I opened my account on Instagram app on my phone and got [this](https://i.postimg.cc/DwkVs2jg/Screenshot.jpg) message.

This 👇 is the part of code which is definitely executed every minute:
```
your_username = '<my_username>'
target_profile = '<target_profile>'
L = instaloader.Instaloader(save_metadata=True, compress_json=False)
L.load_session_from_file(your_username)
P = Profile.from_username(L.context, target_profile)
userID = [P.userid]

for story in L.get_stories(userID):
    story_items = story.get_items()
    ...
    for item in story_items:
        media_id = str(item.mediaid)
        expiry_time = str(item.expiring_local)
        ...
```

And if there has been some change in the target_profile's story only then this 👇 part executes:
```
os.system("instaloader --stories-only --no-posts --no-compress-json --login {0} {1}".format(your_username, target_profile))
```

PS: I'm running the script in background using `pythonw StoryAutomation.py`

I would really appreciate any insight on this. Thanks!

**Comment by Testing2134** [comment]
While not directly related to instaloader, I'm using pyinstalive and making two requests every 10 minutes and that is enough to trip instagrams spam filter or whatever it is. It seems they have tightened up a lot lately. I used to be able to make 10-15 requests per minute.

**Comment by Thammus** [comment]
The HTTP 429 Too Many Requests response status code indicates the user has sent too many requests in a given amount of time ("rate limiting"). (https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/429)

> Please do not run multiple instances of Instaloader [...] within short sequence.

If your script executes every minute, you do run multiple instances of Instaloader within short sequence. Each instance tracks its requests and tries to apply an appropriate rate controlling in order to prevent "429 - Too Many Requests" errors. Invoking new instances resets the tracking of the requests and thus no proper rate controlling logic can be applied. 

> And if there has been some change in the target_profile's story only then this 👇 part executes:
> 
> ```
> os.system("instaloader --stories-only --no-posts --no-compress-json --login {0} {1}".format(your_username, target_profile))
> ```

This is even one more instance of Instaloader that issues untracked requests.

**Comment by Deleora** [comment]
Ahh! Thanks for pointing it out. I'll fix this. Also if I make use of only one instance of Instaloader then what is the safest maximum frequency for running the script without triggering "Max retries exceeded..." exception?

**Comment by Thammus** [comment]
The line
```python
L = instaloader.Instaloader(save_metadata=True, compress_json=False)
```
of your script creates a new Instaloader instance which is stored in the variable `L`. Every time `instaloader.Instaloader()` is executed, a new instance is created.

When Instaloader is used inside a script, each run of the script creates a new instance of Instaloader. Instead of terminating and re-running scripts that use Instaloader, **the safest way is to only start the script once and keep it running** in order to perform certain tasks over time. For example, you would use (and keep) a single Instaloader instance to fetch stories repeatedly and the logic to wait between the subsequent calls of the code fragments that use Instaloader should be part of the script instead of being handled externally.
